File: src/dp/lcs.cpp
// lcs.cpp — CLRS 4e §14.4：LCS 長度表＋回溯重建
#include <iostream>
#include <string>
#include <vector>

std::string lcs(const std::string& X, const std::string& Y) {
    std::size_t m = X.size(), n = Y.size();
    std::vector<std::vector<int>> c(m + 1, std::vector<int>(n + 1, 0));
    for (std::size_t i = 1; i <= m; ++i)
        for (std::size_t j = 1; j <= n; ++j)
            c[i][j] = (X[i-1] == Y[j-1]) ? c[i-1][j-1] + 1                // ↖ 配對
                                         : std::max(c[i-1][j], c[i][j-1]); // ↑ 或 ←
    std::string out;                       // 從 c[m][n] 反向走回去重建
    for (std::size_t i = m, j = n; i > 0 && j > 0;) {
        if (X[i-1] == Y[j-1]) { out = X[i-1] + out; --i; --j; }
        else if (c[i-1][j] >= c[i][j-1]) --i;
        else --j;
    }
    std::cout << "LCS 長度 = " << c[m][n] << '\n';
    return out;
}

int main() {
    // CLRS §14.4 的例子
    std::cout << lcs("ABCBDAB", "BDCABA") << '\n';
    return 0;
}

File: src/dp/lcs.py
# 改寫自官方 Chapter 14/longest_common_subsequence.py（MIT License）：含重建全文
def lcs_length(X, Y):
    m, n = len(X), len(Y)
    c = [[0] * (n + 1) for _ in range(m + 1)]
    b = [[""] * (n + 1) for _ in range(m + 1)]      # 方向表：↖ ↑ ←
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if X[i - 1] == Y[j - 1]:
                c[i][j] = c[i - 1][j - 1] + 1
                b[i][j] = "↖"                       # 配對成功
            elif c[i - 1][j] >= c[i][j - 1]:
                c[i][j] = c[i - 1][j]
                b[i][j] = "↑"
            else:
                c[i][j] = c[i][j - 1]
                b[i][j] = "←"
    return c, b


def print_lcs(b, X, i, j):                          # 沿 ↖ 反向收字
    if i == 0 or j == 0:
        return ""
    if b[i][j] == "↖":
        return print_lcs(b, X, i - 1, j - 1) + X[i - 1]
    if b[i][j] == "↑":
        return print_lcs(b, X, i - 1, j)
    return print_lcs(b, X, i, j - 1)


X, Y = "ABCBDAB", "BDCABA"
c, b = lcs_length(X, Y)
print("LCS 長度 =", c[len(X)][len(Y)])
print("LCS =", print_lcs(b, X, len(X), len(Y)))
